OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,460 @@
+                            uint32_t command_buffer_count, const VkCommandBuffer* command_buffers) {
+    submission_tracker_.UntrackCommandBuffers(device, command_pool, command_buffers,
+                                              command_buffer_count);
+    return dispatch_table_.FreeCommandBuffers(device)(device, command_pool, command_buffer_count,
+                                                      command_buffers);
+  }
+
+  [[nodiscard]] VkResult OnBeginCommandBuffer(VkCommandBuffer command_buffer,
+                                              const VkCommandBufferBeginInfo* begin_info) {
+    VkResult result =
+        dispatch_table_.BeginCommandBuffer(command_buffer)(command_buffer, begin_info);
+    submission_tracker_.MarkCommandBufferBegin(command_buffer);
+    return result;
+  }
+
+  [[nodiscard]] VkResult OnEndCommandBuffer(VkCommandBuffer command_buffer) {
+    submission_tracker_.MarkCommandBufferEnd(command_buffer);
+    return dispatch_table_.EndCommandBuffer(command_buffer)(command_buffer);
+  }
+
+  [[nodiscard]] VkResult OnResetCommandBuffer(VkCommandBuffer command_buffer,
+                                              VkCommandBufferResetFlags flags) {
+    submission_tracker_.ResetCommandBuffer(command_buffer);
+    return dispatch_table_.ResetCommandBuffer(command_buffer)(command_buffer, flags);
+  }
+
+  void OnGetDeviceQueue(VkDevice device, uint32_t queue_family_index, uint32_t queue_index,
+                        VkQueue* queue) {
+    dispatch_table_.GetDeviceQueue(device)(device, queue_family_index, queue_index, queue);
+    queue_manager_.TrackQueue(*queue, device);
+  }
+
+  void OnGetDeviceQueue2(VkDevice device, const VkDeviceQueueInfo2* queue_info, VkQueue* queue) {
+    dispatch_table_.GetDeviceQueue2(device)(device, queue_info, queue);
+    queue_manager_.TrackQueue(*queue, device);
+  }
+
+  [[nodiscard]] VkResult OnQueueSubmit(VkQueue queue, uint32_t submit_count,
+                                       const VkSubmitInfo* submits, VkFence fence) {
+    std::optional<typename SubmissionTracker::QueueSubmission> queue_submission_optional =
+        submission_tracker_.PersistCommandBuffersOnSubmit(submit_count, submits);
+    VkResult result = dispatch_table_.QueueSubmit(queue)(queue, submit_count, submits, fence);
+    submission_tracker_.PersistDebugMarkersOnSubmit(queue, submit_count, submits,
+                                                    queue_submission_optional);
+    return result;
+  }
+
+  [[nodiscard]] VkResult OnQueuePresentKHR(VkQueue queue, const VkPresentInfoKHR* present_info) {
+    submission_tracker_.CompleteSubmits(queue_manager_.GetDeviceOfQueue(queue));
+    return dispatch_table_.QueuePresentKHR(queue)(queue, present_info);
+  }
+
+  void OnCmdBeginDebugUtilsLabelEXT(VkCommandBuffer command_buffer,
+                                    const VkDebugUtilsLabelEXT* label_info) {
+    if (dispatch_table_.IsDebugUtilsExtensionSupported(command_buffer)) {
```

Comment:
Is this "if" needed? Shouldn't we already just fail earlier if this extension is not supported?

Reply:
Same question in a few places below.

Reply:
Ok, sounds good.